romeo/asset-contract/tests/asset_SetBitcoinWalletPublicKeyCommand.ts

```diff
@@ -0,0 +1,61 @@
+import {
+  AssetCommand,
+  Real,
+  Stub,
+  shortenString
+} from "./asset_CommandModel.ts";
+
+import { tx } from "@hirosystems/clarinet-sdk";
+import { Cl } from "@stacks/transactions";
+
+import { expect } from "vitest";
+
+export class SetBitcoinWalletPublicKeyCommand implements AssetCommand {
+  readonly sender: string;
+  readonly pubKey: Uint8Array;
+
+  constructor(
+    sender: string,
+    pubKey: Uint8Array,
+  ) {
+    this.sender = sender;
+    this.pubKey = pubKey;
+  }
+
+  check(_: Readonly<Stub>): boolean {
+    // Can run if sender is the deployer. This is filtered in the generator.
+    return true;
```

Comment:
Instead of comment, it might make sense to enforce the logic in the generator here as well.

Reply:
We should *probably* do it... However, there is a catch when doing that:

Say that the command runs 100 times.

If you enforce the logic there in the `check` call, you'll be [discarding](https://www.cse.chalmers.se/~rjmh/QuickCheck/manual_body.html#6) the test-run if the condition is not met. For example, if the sender is not the `deployer` 30% of the times, you'll end up running the command 30% of the times less.

If you filter in the generator (as-is right now), you guarantee to always run the command with a `deployer` as the sender. I'll double-check if my theory is correct with fast-check though, and will update those `checks` if that's not the case.

---

In general, I'm planning to move us away from those boilerplate-ish commands, and define them on the fly, right next to where the generators are defined. [Technically it seems possible](https://github.com/dubzzz/fast-check/discussions/3026) so all that's left is to try it.

Reply:
I agree with moodmosaic. Furthermore, in my mind, `check` is good for 'managing' model transitions while the generators model input domain.

Reply:
In fast-check, `check` is the place where you decide whether the command (with its generated data) should run or not, given the current state of the model.

---

From the [docs](https://fast-check.dev/docs/advanced/model-based-testing/#define-the-commands):


| Syntax           | Description                                                             |
| ---------------- | ----------------------------------------------------------------------- |
| check(model)     | Ensure that the model is in the appropriate state to execute the action |
| run(model, real) | Execute the action                                                      |
| toString()       | Serialize the command for error reports                                 |